kernel/Utilities/ElementLocalIndexing.cpp

```diff
@@ -0,0 +1,129 @@
+ ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "ElementLocalIndexing.h"
+
+#include "Element.h"
+
+namespace hpgem {
+namespace Utilities {
+
+ElementLocalIndexing::ElementLocalIndexing(std::size_t numberOfUnknowns)
+    : element_(nullptr),
+      numberOfUnknowns_(numberOfUnknowns),
+      totalNumberOfDofs_(0),
+      offsets_(numberOfUnknowns_, -1),
+      sizes_(numberOfUnknowns_, 0),
+      includedUnknowns_(0) {}
+
+void ElementLocalIndexing::reinit(
+    const std::vector<std::size_t> &includedUnknowns) {
+    includedUnknowns_ = includedUnknowns;
+    std::sort(includedUnknowns_.begin(), includedUnknowns_.end());
+#ifdef HPGEM_ASSERTS
+    // Check for duplicates, easy after sorting
+    for (std::size_t i = 0; i < includedUnknowns_.size() - 1; ++i) {
+        logger.assert_debug(includedUnknowns_[i] != includedUnknowns_[i + 1],
+                            "Duplicate unknown");
+    }
+    // Check maximum
+    if (!includedUnknowns_.empty()) {
+        std::size_t maxUnknown =
+            includedUnknowns_[includedUnknowns_.size() - 1];
+        logger.assert_debug(
+            maxUnknown < numberOfUnknowns_,
+            "Largest unknown % is larger than maximum allowable unknown %",
+            maxUnknown, numberOfUnknowns_);
+    }
+#endif
+    // Update the content
+    if (element_ != nullptr) {
+        // Clear the storage to the default values as reinit does only overwrite
+        // the included unknowns.
+        std::fill(offsets_.begin(), offsets_.end(), -1);
+        std::fill(sizes_.begin(), sizes_.end(), 0);
+        reinit(element_);
+    }
+}
+
+void ElementLocalIndexing::reinit(const Base::Element *element) {
+    element_ = element;
+    if (element == nullptr) {
+        std::fill(offsets_.begin(), offsets_.end(), -1);
+        std::fill(sizes_.begin(), sizes_.end(), 0);
+        totalNumberOfDofs_ = 0.0;
+    } else {
+        totalNumberOfDofs_ = 0;
+        for (std::size_t unknown : includedUnknowns_) {
+            std::size_t numDoFs = element->getNumberOfBasisFunctions(unknown);
+            offsets_[unknown] = totalNumberOfDofs_;
+            sizes_[unknown] = numDoFs;
+            totalNumberOfDofs_ += numDoFs;
+        }
+    }
+}
+
+void ElementLocalIndexing::validate() const {
```

Comment:
I am not sure what the idea behind this function is can you turn `validate` into a more descriptive name?

Reply:
Renamed